[smolegz] iP

smolegz opened the pull request:
# *Howie To-Do App*

> Just Do It - Nike

## Here are the reasons why you **definitely** need Howie :ghost::ghost::ghost: :
- Howie is easy-to-use
- Free
- Keeps your life ~~in~~ on track!

## How to use Howie?
1. - [x] Download the latest JAR file from [here](https://github.com/smolegz/ip/releases/tag/A-Jar)!
1.   - [x] Open up your command line.
1.  - [x] Type in the following command:
  ```powershell
  java -jar C:\filepath\of\the\downloaded\jar\file
  ```

## Sample Commands
Adds shower into the to-do list. `todo shower` 

Add a deadline. `deadline Assignment Submission /by Thursday`

Add an event. `event Prom date /from Wed 9pm /to Wed 11pm`

Mark an item as completed. `mark 1`/ `mark 3`

Unmark an item as uncompleted `unmark 3`

To exit. `bye`

Alternatively, you can clone the repository and run Howie.java, as shown below.
```Java
public class Howie {

    public static void main(String[] args) throws Exception {
        Ui ui = new Ui();
        Storage storage = new Storage();
        TaskList taskLs = storage.readFile();
        while (true) {
            String[] input = ui.getUserCommand();
            try {
                Command command = new Parser().parseCommand(input);
                command.setData(taskLs);
                command.execute();
            } catch (DukeException e) {
                System.out.println(e.getMessage());
                Ui.printVLine();
            }
        }
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

smolegz committed:
Completed Level-0

smolegz committed:
Completed Level-1

smolegz committed:
Completed Level-2

smolegz committed:
Completed Level-3

smolegz committed:
Completed Level-4

smolegz committed:
Completed TextUITesting

smolegz committed:
Completed Level-5 and Level-6

smolegz committed:
Added documentation and refined Level-5.

smolegz committed:
Add Level-7

smolegz committed:
Merge branch 'branch-Level-7'

smolegz committed:
Add Level-8

smolegz committed:
Merge branch 'branch-Level-8'

smolegz committed:
Add more OOP

smolegz committed:
Package most files

smolegz committed:
Add JavaDoc and update coding standards

smolegz committed:
Revert "Add JavaDoc and update coding standards"

This reverts commit 05bbfb12c5d6a169eea41be802008e9dd33c3af0.

smolegz committed:
Merge branch 'add-gradle-support'

smolegz committed:
Add tests for TodoCommand and MarkCommand

smolegz committed:
Add Jar

smolegz committed:
Add more JavDoc to several classes.

smolegz committed:
Add find feature.

smolegz committed:
Merge branch 'branch-Level-9'

smolegz committed:
Merge branch 'branch-A-JavaDoc'

smolegz committed:
Merge branch 'master' into branch-A-CodingStandard

smolegz committed:
Add test for FindCommand. Fix some bugs.

smolegz committed:
Configure gradle file.

smolegz committed:
Howie.java : clean up some code

AidenLYT commented on src/main/java/Howie.java:
```diff
@@ -0,0 +1,40 @@
+ * todo, event and deadlines.
+ * @author Koo Zhuo Hui
+ */
+public class Howie {
```
I don't like this name Howie

AidenLYT reviewed (commented):
Test reviewing PR

camille-readbean commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,91 @@
+package parser;
+
+import commands.*;
```
Just a minor note!  
  
Imported classes should always be listed explicitly.  
[coding standard](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements)

smolegz replied:
Thank you for pointing that out. Apparently my IDE does auto-wildcard import when 5 imports are made. I have since disabled that feature :smile:

camille-readbean commented on src/main/java/ui/Ui.java:
```diff
@@ -0,0 +1,106 @@
+     */
+    public static void helpMessage() {
+        printVLine();
+        System.out.println("Below is a list of available commands for me:\n" +
```
Break before line breaks may be better.  
see: https://se-education.org/guides/conventions/java/intermediate.html#layout

camille-readbean commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,49 @@
+
+    @Override
+    public String toString() {
+        return "[D][" + (super.getStatus() ? "X" : " ") + "] " +
```
Break before operators / be consistent in how lines are break
https://se-education.org/guides/conventions/java/intermediate.html#layout

camille-readbean commented on src/main/java/Howie.java:
```diff
@@ -0,0 +1,40 @@
+    public static void main(String[] args) throws Exception {
+        Ui ui = new Ui();
+        Storage storage = new Storage();
+        TaskList taskLs = storage.readFile();
```
`taskLs` may not be an ideal name. Maybe `taskList`.  
I'm not sure if `Ls` is a short form of `List`?

https://se-education.org/guides/conventions/java/intermediate.html#naming

vannykin replied:
Agree with Camille, taskList might be better naming :)

smolegz replied:
Thanks for pointing it out! Would proceed to rectify 👍🏻

camille-readbean commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,91 @@
+        StringBuilder name;
+        StringBuilder current;
+        String s1 = input[0];
+        switch (s1) {
```
I'm not too sure if _switch_ statements can have line breaks between cases.  

https://se-education.org/guides/conventions/java/intermediate.html#naming

camille-readbean commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,91 @@
+        case ByeCommand.COMMAND:
+            return new ByeCommand();
+
+        case MarkCommand.COMMAND:
+            if (input.length != 2) {
+                throw new DukeException("Invalid arguments detected! Please enter a index.");
+            }
+            return new MarkCommand(Integer.parseInt(input[1]));
+
+        case UnmarkCommand.COMMAND:
+            if (input.length != 2) {
+                throw new DukeException("Invalid arguments detected! Please enter a index.");
+            }
+            return new UnmarkCommand(Integer.parseInt(input[1]));
+
+        case TodoCommand.COMMAND:
+            name = new StringBuilder();
+            for (int j = 1; j < input.length; j++) {
+                name.append(input[j]).append(" ");
+            }
+            return new TodoCommand(name.toString());
```
Would love to hear your rational of why you decided to place the parameters extraction inside the switch statements! 😄

smolegz replied:
Initially, I was also unsure if whether parameters should be extracted within the Parser class, or during the execution phase. However, I decided that the execution phase should only be executed when there's a valid input given by the user; any wrong input would be rejected by the parser. I hope that answers ur question... do lmk if you have any recommendations or any queries. Thx!

camille-readbean commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,112 @@
+                        }
+                        tasks.add(t);
+                    } else if (input[0].equals("D")) {
+                        Deadline d = new Deadline(input[2],input[3]);
+                        if (input[1].equals("1")) {
+                            d.setDone();
+                        }
+                        tasks.add(d);
+                    } else {
+                        Event e = new Event(input[2], input[3], input[4]);
+                        if (input[1].equals("1")) {
+                            e.setDone();
+                        }
+                        tasks.add(e);
+                    }
+                }
+            }
+            return tasks;
+        } catch (IOException e) {
+            throw new DukeException("I/O error detected...");
+        }
+    }
+
+    public void saveFile(TaskList taskLs) throws DukeException, IOException {
+        List<Task> tasks = taskLs.getList();
+        try {
+            BufferedWriter bw = new BufferedWriter(new FileWriter(path.toString()));
+            for (Task t : tasks) {
+                bw.write(t.encode());
+                bw.newLine();
+            }
+            bw.flush();
+            bw.close();
+        } catch (FileNotFoundException e) {
+           Files.createDirectories(Paths.get("data"));
+           saveFile(taskLs);
+        } catch (IOException e) {
+            throw new DukeException("I/O Exception Detected");
+        }
+    }
```
I like the way you did it 👍 
How it supports multiple formats for an example.

camille-readbean reviewed (commented):
Looks good to me 👍   
  
Definitely can learn a lot! I like how you did the OOP!

smolegz committed:
Update README.md

smolegz committed:
Update README.md

vannykin commented on src/main/java/parser/Parser.java:
```diff
@@ -21,9 +22,15 @@ public Command parseCommand(String[] input) {
             return new ByeCommand();
```
Not sure why there is an extra indentation for 'return new ByeCommand()' here.
Minor typo probably!

smolegz replied:
Amended!

vannykin reviewed (commented):
Good job :) None of the names are problematic to me, clean and clear code yay

smolegz committed:
Improve coding standards

smolegz committed:
Merge branch 'master' of https://github.com/smolegz/ip

smolegz committed:
Implement JavaFx GUI

smolegz committed:
Package Howie.java and GUI files

smolegz committed:
Fix some styles to GUI

smolegz committed:
Add and fix GUI features

smolegz committed:
Test PR

smolegz committed:
Testing PR

smolegz committed:
Add several assertion statements

There is a possibility that commands execution are not be behaving
as intended.

Several assertion statements have been included to guard against any
potential errors. Particularly, assertion statements have been added
to TodoCommand.java, Deadline.java and Event.java etc.

These checks ensures that correct task names are added into
TaskList.

smolegz committed:
Merge pull request #1 from smolegz/branch-A-Assertions

Add Assertions to Check for Errors

smolegz committed:
Refactor existing codebase

The code quality can be further improved to increase readability.

Some of the refactoring consisted of replacing magic literal,
extracting methods and replacing nested conditionals with guard
clauses.

This is done to so that code can be easily maintained.

smolegz committed:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	data/data.txt
#	src/main/java/commands/DeadlineCommand.java
#	src/main/java/commands/EventCommand.java
#	src/main/java/commands/FindCommand.java
#	src/main/java/commands/MarkCommand.java
#	src/main/java/commands/TodoCommand.java
#	src/main/java/commands/UnmarkCommand.java
#	src/main/java/howie/Howie.java
#	src/main/java/storage/Storage.java

smolegz committed:
Fix code

smolegz committed:
Add C-Help

smolegz committed:
Add UserGuide and some touch-ups

smolegz committed:
Update userguide

smolegz committed:
Update README.md

smolegz committed:
Refactor Parser class

smolegz committed:
Merge branch 'master' of https://github.com/smolegz/ip

# Conflicts:
#	docs/README.md

smolegz committed:
Fix JUnit Test

smolegz committed:
Add more JavaDoc
